src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -1311,3 +1329,86 @@ double GetStreamCpuLoad( PaStream * s )
 
     return PaUtil_GetCpuLoad( &stream->cpuLoadMeasurer );
 }
+
+/** Extensions */
+static void RenameStreamCb(pa_stream *s, int success, void *userdata)
+{
+    /* Currently does nothing but signal the caller. */
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) userdata;
+    pa_threaded_mainloop_signal( l_ptrStream->mainloop,
+                                 0 );
+}
+
+PaError PaPulseAudio_RenameSource( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+
+    if ( stream->inStream == NULL )
+    {
+        return paInvalidDevice;
+    }
+
+    /* Reallocate stream name in memory. */
+    pa_threaded_mainloop_lock( stream->mainloop );
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    if ( !newStreamName )
+    {
+        pa_threaded_mainloop_unlock( stream->mainloop );
+        return paInsufficientMemory;
+    }
+    memset(newStreamName, 0, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    strncpy(newStreamName, streamName, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME));
+
+    PaUtil_FreeMemory( stream->sourceStreamName );
+    stream->sourceStreamName = newStreamName;
+
+    op = pa_stream_set_name( stream->inStream, streamName, RenameStreamCb, stream );
+    pa_threaded_mainloop_unlock( stream->mainloop );
+
+    /* Wait for completion. */
+    while (pa_operation_get_state( op ) == PA_OPERATION_RUNNING)
+    {
+        pa_threaded_mainloop_wait( stream->mainloop );
+    }
+
+    return result;
+}
+
+PaError PaPulseAudio_RenameSink( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+    
+    if ( stream->outStream == NULL )
+    {
+        return paInvalidDevice;
+    }
+
+    /* Reallocate stream name in memory. */
+    pa_threaded_mainloop_lock( stream->mainloop );
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    if ( !newStreamName )
+    {
+        pa_threaded_mainloop_unlock( stream->mainloop );
+        return paInsufficientMemory;
+    }
+    memset(newStreamName, 0, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    strncpy(newStreamName, streamName, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME));
+
+    PaUtil_FreeMemory( stream->sinkStreamName );
+    stream->sinkStreamName = newStreamName;
+    
+    op = pa_stream_set_name( stream->outStream, streamName, RenameStreamCb, stream );
```

Comment:
does it work if stream->outStream is still null?

Reply:
Line 1385 has a NULL check for stream->outStream and returns paInvalidDevice if true.

Reply:
We  have little chicken and egg problem here and below as stream is created in initialization.. so actually outStream or inStream doesn't exists before initialization. So we are in coders good will that renaming is called after that. I think we just mention this in Documentation as kind of warning and trust that people read that.

Reply:
For my understanding it should work.

As soon we have a "stream" which is a Portaudio structure we can add a name to stream->sinkStreamName. 

If we have a stream->outStream, at that time, we need to update that as well by calling the pulse function. If not, it is not necessary, because the code that opens the stream will initialize it wit the stored name.

Reply:
This will work only after OpenStream before that there is no allocated pointer to PaStream and you can't store anything.

Reply:
Ups... 
So the whole idea is wrong to set the stream name first. We cannot do it, because we can't store it before, right?

Reply:
@dashuer you are correct. This works only with streams that are already running. Thats why to have name for stream we need to to use PaStreamParameters->hostApiSpecificStreamInfo with some struct which have name and other needed stuff

Reply:
For what it's worth, I haven't noticed any issues with stream renames not immediately taking effect with the current code, so it's possible that the function calls only working after Pa_OpenStream()/Pa_StartStream() is fine. Support for hostApiSpecificStreamInfo could be put into place in a future PR if desired.

Of course, it's also okay to implement it prior to merging the parent PA if you'd like :)

Reply:
@tmiw Would it be easiest to get PR without this in and then start uphill battle as I know this will be personally I see this LGTM state.

Reply:
@illuusio, the current implementation seems close to ready so it might be easiest to proceed with this. However, if the other PortAudio maintainers prefer hostApiSpecificStreamInfo (in addition to or instead of this), that's a different story. If that's the case, I'm fine with deferring this and getting basic PortAudio support in.